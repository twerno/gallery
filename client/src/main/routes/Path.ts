import RouteUtils from 'main/utils/RouteUtils';

export interface IGalleryUrlQuery {
    q: string | undefined;
}

export const Path = {
    homeRoute: '/',
    homeUrl() { return this.homeRoute },

    galleryRoute: '/gallery/search',
    galleryUrl(query: IGalleryUrlQuery) { return RouteUtils.makeUrl(this.galleryRoute, null, query) },

};
