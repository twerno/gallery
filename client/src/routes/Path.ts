import RouteUtils from 'utils/RouteUtils';

export const Path = {
    homeRoute: '/',
    homeUrl() { return this.homeRoute },

    galleryRoute: '/gallery/search/:query',
    galleryUrl(query: string) { return RouteUtils.makeUrl(this.galleryRoute, { query }) },

};
