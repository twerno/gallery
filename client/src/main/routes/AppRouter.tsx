import AnimatedLoader from 'main/components/AnimatedLoader';
import FullScreenContainer from 'main/components/FullScreenContainer';
import * as React from 'react';
import { Route, Switch } from 'react-router';
import { HomePage } from './home/HomePage';
import { Path } from './Path';


const LazyGalleryPage = React.lazy(() => import(/* webpackChunkName: "GalleryPage" */ './gallery/GalleryPage'));

export interface IAppRouterProps {

}

export const AppRouter: React.FC<IAppRouterProps> = (props) => {
    return (
        <React.Suspense fallback={
            <FullScreenContainer position='absolute'>
                <AnimatedLoader />
            </FullScreenContainer>
        }>
            <Switch>
                <Route
                    path={Path.galleryRoute}
                    render={(routeProps) => <LazyGalleryPage routeProps={routeProps} />}
                />
                <Route
                    path={Path.homeRoute}
                    exact={true}
                    render={() => <HomePage />}
                />
                <Route render={() => <div>404</div>} />
            </Switch>
        </React.Suspense>
    );
}