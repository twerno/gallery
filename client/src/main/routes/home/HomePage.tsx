import { Path } from 'main/routes/Path';
import * as React from 'react';
import { useHistory } from 'react-router-dom';
import HomeContainer from './components/styled/HomeContainer';
import HomePageSearchForm from './components/styled/HomePageSearchForm';
import { IGalleryUrlQuery, GalleryQueryHelper } from '../gallery/model/galleryQuery';

export interface IHomePageProps {
    initialQuery?: IGalleryUrlQuery;
}

export const HomePage = ({ initialQuery }: IHomePageProps) => {

    const history = useHistory();

    const onSubmit = async (values: IGalleryUrlQuery) => {
        const newPath = Path.galleryUrl({ q: values.q });
        history.push(newPath);
    }

    return (
        <HomeContainer>
            <h1>React Gallery</h1>
            <HomePageSearchForm
                initialValues={initialQuery || GalleryQueryHelper.buildEmpty()}
                onSubmit={onSubmit}
            />
        </HomeContainer>
    );
}