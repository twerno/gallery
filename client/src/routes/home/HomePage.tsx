import Button from 'components/styled/Button';
import Input from 'components/styled/Input';
import { Field, Formik } from 'formik';
import * as React from 'react';
import { useHistory } from 'react-router-dom';
import { Path } from 'routes/Path';

import HomeContainer from './components/styled/HomeContainer';
import HomeSearchForm from './components/styled/HomeSearchForm';

export interface IHomePageProps {
}

export const HomePage = (props: IHomePageProps) => {

    const history = useHistory();

    return (
        <HomeContainer>
            <h1>React Gallery</h1>
            <Formik
                initialValues={{ q: '' }}
                onSubmit={(values) => {
                    history.push(Path.galleryUrl({ q: values.q }));
                }}
            >
                {
                    ({ isSubmitting, handleSubmit, handleReset }) => (
                        <HomeSearchForm>
                            <Field name="q" as={Input} aria-label="Search" />
                            <Button type="submit" disabled={isSubmitting}>Search</Button>
                        </HomeSearchForm>
                    )
                }
            </Formik>
        </HomeContainer>
    );
}