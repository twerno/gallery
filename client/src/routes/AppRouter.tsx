import * as React from 'react';
import { Route, Switch } from 'react-router';

import { GalleryPage } from './gallery/GalleryPage';
import { HomePage } from './home/HomePage';
import { Path } from './Path';

export interface IAppRouterProps {

}

export const AppRouter: React.FC<IAppRouterProps> = (props) => {
    return (
        <Switch>
            <Route
                path={Path.galleryRoute}
                render={(routeProps) => <GalleryPage routeProps={routeProps} />}
            />
            <Route
                path={Path.homeRoute}
                exact={true}
                render={() => <HomePage />}
            />
            <Route render={() => <div>404</div>} />
        </Switch>
    );
}