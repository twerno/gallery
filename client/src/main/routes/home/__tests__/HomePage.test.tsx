import '@testing-library/jest-dom/extend-expect';
import { wait } from '@testing-library/react';
import { mount } from 'enzyme';
import 'jest-styled-components';
import * as React from 'react';
import { componentBasicTests } from 'test/componentBasicTests';
import { HomePage } from '../HomePage';
import { Path } from 'main/routes/Path';

const historyMock = {
    location: {},
    push: jest.fn(),
};

jest.mock('react-router-dom', () => ({
    useHistory: () => historyMock,
}));

describe('HomePage', () => {

    beforeEach(() => {
        historyMock.push.mockClear();
    })

    componentBasicTests("HomePage", <HomePage />);
    componentBasicTests("HomePage", <HomePage initialQuery={{ q: 'cats' }} />);

    test('searching for cats', async () => {

        const wrapper = mount(
            <HomePage initialQuery={{ q: 'cats' }} />
        );

        await wait(() => wrapper.find('form').simulate('submit'));

        expect(historyMock.push).toBeCalledWith(Path.galleryUrl({ q: 'cats' }));
    });

    test('searching for dogs', async () => {

        const wrapper = mount(
            <HomePage initialQuery={{ q: 'dogs' }} />
        );

        await wait(() => wrapper.find('form').simulate('submit'));

        expect(historyMock.push).toBeCalledWith(Path.galleryUrl({ q: 'dogs' }));
        expect(historyMock.push).not.toBeCalledWith(Path.galleryUrl({ q: 'cats' }));
    });

});

