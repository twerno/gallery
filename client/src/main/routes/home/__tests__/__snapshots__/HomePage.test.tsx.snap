// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage [HomePage] Component basic tests matches the snapshot 1`] = `
<styled.div>
  <h1>
    React Gallery
  </h1>
  <Component
    initialValues={
      Object {
        "q": undefined,
      }
    }
    onSubmit={[Function]}
  />
</styled.div>
`;

exports[`HomePage [HomePage] Component basic tests matches the snapshot 2`] = `
<styled.div>
  <h1>
    React Gallery
  </h1>
  <Component
    initialValues={
      Object {
        "q": "cats",
      }
    }
    onSubmit={[Function]}
  />
</styled.div>
`;
